src/lib/core/authorization/AuthorizationManager.ts

```diff
@@ -0,0 +1,375 @@
+    let handler = () => {};
+    if (isAuthorizationRequirementsError(response)) {
+      log(
+        'debug',
+        'AuthorizationManager.handleErrorResponse | error=AuthorizationRequirementsError',
+      );
+      handler = () => this.handleAuthorizationRequirementsError(response);
+    }
+    if (isConsentRequiredError(response)) {
+      log('debug', 'AuthorizationManager.handleErrorResponse | error=ConsentRequiredError');
+      handler = () => this.handleConsentRequiredError(response);
+    }
+    if ('code' in response && response['code'] === 'AuthenticationFailed') {
+      log('debug', 'AuthorizationManager.handleErrorResponse | error=AuthenticationFailed');
+      handler = () => this.revoke();
+    }
+    return execute === true ? handler() : handler;
+  }
+
+  /**
+   * Process a well-formed Authorization Requirements error response from a Globus service
+   * and redirect the user to the Globus Auth login page with the necessary parameters.
+   */
+  handleAuthorizationRequirementsError(response: AuthorizationRequirementsError) {
+    this.#transport = this.#buildTransport({
+      params: {
+        session_message: response.authorization_parameters.session_message,
+        session_required_identities:
+          response.authorization_parameters.session_required_identities.join(','),
+        session_required_mfa: response.authorization_parameters.session_required_mfa,
+        session_required_single_domain:
+          response.authorization_parameters.session_required_single_domain.join(','),
+        prompt: 'login',
+      },
+    });
+    this.#transport.send();
+  }
+
+  /**
+   * Process a well-formed `ConsentRequired` error response from a Globus service
+   * and redirect the user to the Globus Auth login page with the necessary parameters.
+   */
+  handleConsentRequiredError(response: ConsentRequiredError) {
+    this.#transport = this.#buildTransport({
+      requested_scopes: this.#withOfflineAccess(response.required_scopes.join(' ')),
+    });
+    this.#transport.send();
+  }
+
+  /**
+   * Add a Globus Auth token response to storage, if `other_tokens` are present they are also added.
+   * This method is mostly used internally by the `AuthorizationManager`, but can be used by downstream
+   * consumers to add tokens to storage if necessary.
+   */
+  addTokenResponse = (token: Token | TokenResponse) => {
+    getStorage().set(`${this.configuration.client}:${token.resource_server}`, token);
+    if ('other_tokens' in token) {
+      token.other_tokens?.forEach(this.addTokenResponse);
+    }
+    this.#checkAuthorizationState();
+  };
+
+  /**
+   * Call `AuthroizationManager.reset`, revoke all of the available tokns, and emit the `revoke` event.
+   * @emits AuthorizationManager.events#revoke
+   * @see AuthorizationManager.reset
+   */
+  async revoke() {
+    log('debug', 'AuthorizationManager.revoke');
+    const revocation = Promise.all(this.tokens.getAll().map((token) => this.#revokeToken(token)));
```

Comment:
nit: can this just be `.map(this.#revokeToken)`?

Reply:
`Promise.all(this.tokens.getAll().map(this.#revokeToken.bind(this)));` is the best I can do based on the context.